chat-plugins/modlog.js

```diff
@@ -145,12 +137,70 @@ function getMoreButton(room, search, useExactSearch, lines, maxLines) {
 	if (!newLines || lines < maxLines) {
 		return ''; // don't show a button if no more pre-set increments are valid or if the amount of results is already below the max
 	} else {
-		if (useExactSearch) search = Chat.escapeHTML(`"${search}"`);
+		search = Chat.escapeHTML(`${search}`);
 		return `<br /><div style="float:right"><button class="button" name="send" value="/modlog ${room}, ${search} ${LINES_SEPARATOR}${newLines}" title="View more results">More results...</button></div>`;
 	}
 }
 
-async function runModlog(rooms, searchString, exactSearch, maxLines) {
+function toModlogId(text) {
+	if (typeof text !== 'string' && typeof text !== 'number') return '';
+	return ('' + text).toLowerCase().replace(/[^a-z0-9*]+/g, '').replace(/\*/g, '.*').replace(/[a-z0-9]/g, '$&[^a-zA-Z0-9]*');
+}
+
+function escapeRegex(str) {
+	return (str && typeof str === 'string' ? str : '').replace(/[\\.+*?!=()|[\]{}^$#<>]/g, '\\$&');
+}
+
+function formatSingleSearch(string) {
+	const searchType = string.trim().charAt(0) === '!' ? 'negative' : 'positive';
+	const parts = string.split('||');
+	let formattedParts = [];
+
+	for (let p of parts) {
+		p = p.trim();
+		const exactSearch = /^(["']).+\1$/i.test(p);
+		if (exactSearch) {
+			p = p.slice(1, -1);
+			formattedParts.push(escapeRegex(p));
+		} else {
+			formattedParts.push(`[^a-zA-Z0-9]${toModlogId(p)}[^a-zA-Z0-9]`);
+		}
+	}
+
+	const result = {searchType, term: (formattedParts.length > 1 ? `(${formattedParts.join('|')})` : formattedParts[0])};
+	return result;
+}
+
+function formatRegexString(string) {
+	const parts = string.split('&&');
+	if (parts.length === 1 && string.charAt(0) !== '!' && !string.includes('||')) {
+		// single argument searches
+		if (/^(["']).+\1$/i.test(string)) {
+			// exact search
+			return {type: 1, searchRegex: escapeRegex(string.slice(1, -1))};
+		} else {
+			// userid search
+			return {type: 0, searchRegex: `[^a-zA-Z0-9]${toModlogId(string)}[^a-zA-Z0-9]`};
+		}
+	} else {
+		// searches with multiple arguments
+		let results = {positive: [], negative: []};
+
+		// format each search term
+		for (let p of parts) {
+			const res = formatSingleSearch(p);
+			results[res.searchType].push(res.term);
+		}
+
+		// format the positive searches
+		const positive = results.positive.map(t => `(?=.*${t})`).join('');
+		const negative = results.negative.length ? results.negative.map(t => `(?!${results.negative.join('|')})`) : '';
+		const searchRegex = `^${positive}(?:${negative}.)*$`;
```

Comment:
All these lookaheads make me nervous.
Someone review?

Reply:
I understand that lookaheads probably aren't the fastest, which is why they're avoided like the devil when the logic operators aren't used, but that's the only way to make sure that 2 or more elements will exist in the same string, as far as I've learned.

Maybe we could limit to say 3 ``&&`` separators so there's only a maximum possibility of 4 lookaheads?